File: MySDKs/JAVA SDK/src/main/java/com/example/sdk/internal/SDKContext.java
package com.example.sdk.internal;

import com.example.sdk.SDKException;
import com.google.gson.Gson;

import javax.crypto.SecretKey;
import javax.crypto.spec.SecretKeySpec;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.locks.ReentrantLock;

/**
 * Manages the internal state of the SDK, including configuration,
 * session data, and a thread-safe key cache.
 */
public class SDKContext {
    private volatile boolean initialized = false;
    private String apiEndpoint;
    private String jwt;
    private final Map<String, SecretKey> keyCache = new ConcurrentHashMap<>();
    private final ReentrantLock sessionLock = new ReentrantLock();
    private final NetClient netClient;

    public SDKContext() {
        this.netClient = new NetClient();
    }

    public void initialize(String regToken) throws SDKException {
        if (regToken == null || regToken.isEmpty()) {
            throw new SDKException("Registration token cannot be null or empty.");
        }

        // Mock backend endpoint
        this.apiEndpoint = "https://api.example-crypto.com/v1";

        // Fetch and parse configuration
        String configJson = netClient.fetchConfig(regToken, this.apiEndpoint);
        // Using Gson to parse config. In a real app, you'd map this to a config object.
        Gson gson = new Gson();
        Map<?, ?> configMap = gson.fromJson(configJson, Map.class);
        System.out.println("SDK_Init: Configuration fetched: " + configMap);

        this.initialized = true;
    }

    public String createSession(String identity, String secret) throws SDKException {
        sessionLock.lock();
        try {
            this.jwt = netClient.authenticate(identity, secret, this.apiEndpoint);
            return this.jwt;
        } finally {
            sessionLock.unlock();
        }
    }

    public void keyOperation(String opType, String keyName, byte[] keyData) throws SDKException {
        if (this.jwt == null) {
            throw new SDKException("No active session. Please call createSession() first.");
        }

        String response = netClient.keyOp(this.jwt, opType, keyName, keyData, this.apiEndpoint);

        if ("READ".equalsIgnoreCase(opType)) {
            // Assuming the response is the raw key material
            byte[] rawKey = response.getBytes(); // In a real scenario, this would be Base64 decoded
            SecretKey secretKey = new SecretKeySpec(rawKey, 0, rawKey.length, "AES");
            keyCache.put(keyName, secretKey);
        } else if ("DELETE".equalsIgnoreCase(opType)) {
            keyCache.remove(keyName);
        }
    }

    public SecretKey getKey(String keyName) throws SDKException {
        // Check cache first
        SecretKey key = keyCache.get(keyName);
        if (key != null) {
            return key;
        }

        // If not in cache, fetch from backend
        System.out.println("Key '" + keyName + "' not in cache. Fetching from backend...");
        keyOperation("READ", keyName, null);

        // Try getting from cache again
        key = keyCache.get(keyName);
        if (key == null) {
            throw new SDKException("Key not found: " + keyName);
        }
        return key;
    }

    public boolean isInitialized() {
        return initialized;
    }

    public void cleanup() {
        // Securely clear sensitive data
        this.jwt = null;
        keyCache.clear();
        this.initialized = false;
    }
}

File: MySDKs/NetCore SDK/Internal/SDKContext.cs
using System.Collections.Concurrent;
using System.Text.Json;

namespace MyCompany.CryptoSDK.Internal;

/// <summary>
/// Manages the internal state of the SDK, including configuration,
/// session data, and a thread-safe key cache.
/// </summary>
internal class SDKContext : IDisposable
{
    private string? _apiEndpoint;
    private string? _jwt;
    private readonly ConcurrentDictionary<string, byte[]> _keyCache = new();
    private readonly NetClient _netClient;

    public bool IsInitialized { get; private set; }

    public SDKContext()
    {
        _netClient = new NetClient();
    }

    public async Task InitializeAsync(string regToken)
    {
        if (string.IsNullOrEmpty(regToken))
        {
            throw new SDKException("Registration token cannot be null or empty.");
        }

        // Mock backend endpoint
        _apiEndpoint = "https://api.example-crypto.com/v1";

        // Fetch and parse configuration
        string configJson = await _netClient.FetchConfigAsync(regToken, _apiEndpoint);
        
        // Using System.Text.Json to parse config.
        var configMap = JsonSerializer.Deserialize<Dictionary<string, object>>(configJson);
        Console.WriteLine($"SDK_Init: Configuration fetched: {configJson}");

        IsInitialized = true;
    }

    public async Task<string> CreateSessionAsync(string identity, string secret)
    {
        _jwt = await _netClient.AuthenticateAsync(identity, secret, _apiEndpoint!);
        return _jwt;
    }

    public async Task KeyOperationAsync(string opType, string keyName, byte[]? keyData)
    {
        if (string.IsNullOrEmpty(_jwt))
        {
            throw new SDKException("No active session. Please call CreateSessionAsync() first.");
        }

        string response = await _netClient.KeyOpAsync(_jwt, opType, keyName, keyData, _apiEndpoint!);

        if ("READ".Equals(opType, StringComparison.OrdinalIgnoreCase))
        {
            // Assuming the response is the raw key material
            byte[] rawKey = Convert.FromHexString(response); // Use Hex in .NET 6+
            _keyCache.TryAdd(keyName, rawKey);
        }
        else if ("DELETE".Equals(opType, StringComparison.OrdinalIgnoreCase))
        {
            _keyCache.TryRemove(keyName, out _);
        }
    }

    public async Task<byte[]> GetKeyAsync(string keyName)
    {
        // Check cache first
        if (_keyCache.TryGetValue(keyName, out byte[]? key))
        {
            return key;
        }

        // If not in cache, fetch from backend
        Console.WriteLine($"Key '{keyName}' not in cache. Fetching from backend...");
        await KeyOperationAsync("READ", keyName, null);

        // Try getting from cache again
        if (_keyCache.TryGetValue(keyName, out byte[]? fetchedKey))
        {
            return fetchedKey;
        }

        throw new SDKException($"Key not found: {keyName}");
    }

    public void Cleanup()
    {
        // Securely clear sensitive data
        _jwt = null;
        foreach (var entry in _keyCache)
        {
            Array.Clear(entry.Value, 0, entry.Value.Length);
        }
        _keyCache.Clear();
        IsInitialized = false;
    }

    public void Dispose()
    {
        Cleanup();
        _netClient.Dispose();
    }
}
